node/node.go

```diff
@@ -350,154 +313,273 @@ func (node *Node) AddPendingReceipts(receipts *types.CXReceiptsProof) {
+	}
+	senderKey = key
+
+	if err := consensus.VerifyMessageSig(senderKey, &m); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	return &m, nil
+}
+
 // Start kicks off the node message handling
 func (node *Node) Start() error {
-	allTopics := node.host.AllTopics()
-	if len(allTopics) == 0 {
-		return errors.New("have no topics to listen to")
+	type t struct {
+		tp    nodeconfig.GroupID
+		isCon bool
+	}
+	groups := map[nodeconfig.GroupID]bool{}
+	for _, t := range []t{
+		{node.NodeConfig.GetShardGroupID(), true},
+		{nodeconfig.NewClientGroupIDByShardID(shard.BeaconChainShardID), false},
+		{node.NodeConfig.GetClientGroupID(), false},
+	} {
+		if _, ok := groups[t.tp]; !ok {
+			groups[t.tp] = t.isCon
+		}
 	}
 
-	const (
-		maxMessageHandlers = 200
-		threshold          = maxMessageHandlers / 2
-		lastLine           = 20
-		throttle           = 100 * time.Millisecond
-		emrgThrottle       = 250 * time.Millisecond
-	)
+	type u struct {
+		p2p.NamedTopic
+		consensusBound bool
+	}
+
+	var allTopics []u
+
+	utils.Logger().Debug().
+		Interface("topics-ended-up-with", groups).
+		Uint32("shard-id", node.Consensus.ShardID).
+		Msg("starting with these topics")
+
+	for key, isCon := range groups {
+		topicHandle, err := node.host.GetOrJoin(string(key))
+		if err != nil {
+			return err
+		}
+		allTopics = append(
+			allTopics, u{
+				NamedTopic:     p2p.NamedTopic{string(key), topicHandle},
+				consensusBound: isCon,
+			},
+		)
+	}
 
+	errMsgHadNoHMYPayLoadAssumption := errors.New("did not have sufficient size for hmy msg")
+	errConsensusMessageOnUnexpectedTopic := errors.New("received consensus on wrong topic")
+
+	pubsub := node.host.PubSub()
 	ownID := node.host.GetID()
-	errChan := make(chan error)
+	errChan := make(chan withError)
+
+	type p2pHandlerConsensus func(
+		ctx context.Context,
+		msg *msg_pb.Message,
+	) error
+
+	type p2pHandlerElse func(
+		ctx context.Context,
+		rlpPayload []byte,
+	) error
+
+	type validated struct {
+		consensusBound bool
+		handleC        p2pHandlerConsensus
+		handleCArg     *msg_pb.Message
+		handleE        p2pHandlerElse
+		handleEArg     []byte
+	}
+
+	isThisNodeAnExplorerNode := node.NodeConfig.Role() == nodeconfig.ExplorerNode
 
 	for i := range allTopics {
 		sub, err := allTopics[i].Topic.Subscribe()
 		if err != nil {
 			return err
 		}
+
 		topicNamed := allTopics[i].Name
-		sem := semaphore.NewWeighted(maxMessageHandlers)
-		msgChan := make(chan *libp2p_pubsub.Message)
-		needThrottle, releaseThrottle :=
-			make(chan time.Duration), make(chan struct{})
+		isConsensusBound := allTopics[i].consensusBound
+
+		utils.Logger().Info().
+			Str("topic", topicNamed).
+			Msg("enabled topic validation on consensus bound messages")
+
+		if err := pubsub.RegisterTopicValidator(
+			topicNamed,
+			func(ctx context.Context, peer libp2p_peer.ID, msg *libp2p_pubsub.Message) bool {
+				hmyMsg := msg.GetData()
+				if len(hmyMsg) < p2pMsgPrefixSize {
+					pubsub.BlacklistPeer(peer)
```

Comment:
is it possible that valid peers are blacklisted by mistake? How long the blacklisting last?

Reply:
In principle no - because harmony protocol respecting nodes will never do this. 

What time period you'd like for the blacklisting to last?

Reply:
If it's mostly for spamming. 10 mins should be enough.

Reply:
k, I set it to 15 already in pushed commit , 10 or 15?

Reply:
actually commented on this - the time based one is bad to use https://github.com/libp2p/go-libp2p-pubsub/issues/338